** WARNING: connection is not using a post-quantum key exchange algorithm.
** This session may be vulnerable to "store now, decrypt later" attacks.
** The server may need to be upgraded. See https://openssh.com/pq.html
diff --git a/node_modules/mineflayer/lib/plugins/digging.js b/node_modules/mineflayer/lib/plugins/digging.js
index 496cf63..813a667 100644
--- a/node_modules/mineflayer/lib/plugins/digging.js
+++ b/node_modules/mineflayer/lib/plugins/digging.js
@@ -242,11 +242,14 @@ function inject (bot) {
       enchantments = enchantments.concat(helmetEnchantments)
     }
 
+    if (block.material === 'incorrect_for_wooden_tool')
+      block.material = 'mineable/pickaxe'
+
     const creative = bot.game.gameMode === 'creative'
     return block.digTime(
       type,
       creative,
-      bot.entity.isInWater,
+      false, // isInWater, water calc is bugged, so always false
       !bot.entity.onGround,
       enchantments,
       bot.entity.effects
diff --git a/node_modules/mineflayer/lib/plugins/inventory.js b/node_modules/mineflayer/lib/plugins/inventory.js
index 70db312..56ee22f 100644
--- a/node_modules/mineflayer/lib/plugins/inventory.js
+++ b/node_modules/mineflayer/lib/plugins/inventory.js
@@ -125,10 +125,15 @@ function inject (bot, { hideErrors }) {
         cursorZ: -1
       })
     } else if (bot.supportFeature('useItemWithOwnPacket')) {
+      const yawDeg = -bot.entity.yaw * 180 / Math.PI + 180 // invert + shift
+      const pitchDeg = -bot.entity.pitch * 180 / Math.PI // invert
       bot._client.write('use_item', {
         hand: offHand ? 1 : 0,
         sequence,
-        rotation: { x: 0, y: 0 }
+        rotation: {
+          x: yawDeg,
+          y: pitchDeg
+        }
       })
     }
   }
@@ -673,6 +678,12 @@ function inject (bot, { hideErrors }) {
   })
 
   function prepareWindow (window) {
+    // [mindaxis-patch:merchant-window-open] Merchant windows may not send window_items; emit windowOpen immediately
+    if (window.type === 'minecraft:merchant' || window.type === 'minecraft:villager') {
+      extendWindow(window)
+      bot.emit('windowOpen', window)
+      return
+    }
     if (!windowItems || window.id !== windowItems.windowId) {
       // don't emit windowOpen until we have the slot data
       bot.once(`setWindowItems:${window.id}`, () => {
diff --git a/node_modules/mineflayer/lib/plugins/place_block.js b/node_modules/mineflayer/lib/plugins/place_block.js
index fdaec6b..08983b6 100644
--- a/node_modules/mineflayer/lib/plugins/place_block.js
+++ b/node_modules/mineflayer/lib/plugins/place_block.js
@@ -11,7 +11,7 @@ function inject (bot) {
     let newBlock = bot.blockAt(dest)
     if (oldBlock.type === newBlock.type) {
       [oldBlock, newBlock] = await onceWithCleanup(bot, `blockUpdate:${dest}`, {
-        timeout: 5000,
+        timeout: 500,
         // Condition to wait to receive block update actually changing the block type, in case the bot receives block updates with no changes
         // oldBlock and newBlock will both be null when the world unloads
         checkCondition: (oldBlock, newBlock) => !oldBlock || !newBlock || oldBlock.type !== newBlock.type
diff --git a/node_modules/mineflayer/lib/plugins/place_entity.js b/node_modules/mineflayer/lib/plugins/place_entity.js
index c7b750a..3abdfad 100644
--- a/node_modules/mineflayer/lib/plugins/place_entity.js
+++ b/node_modules/mineflayer/lib/plugins/place_entity.js
@@ -36,9 +36,7 @@ function inject (bot) {
 
     if (type === 'boat') {
       if (bot.supportFeature('useItemWithOwnPacket')) {
-        bot._client.write('use_item', {
-          hand: options.offhand ? 1 : 0
-        })
+        bot.activateItem(options.offhand)
       } else {
         bot._client.write('block_place', {
           location: { x: -1, y: -1, z: -1 },
diff --git a/node_modules/mineflayer/lib/plugins/villager.js b/node_modules/mineflayer/lib/plugins/villager.js
index e20e34a..59a949a 100644
--- a/node_modules/mineflayer/lib/plugins/villager.js
+++ b/node_modules/mineflayer/lib/plugins/villager.js
@@ -76,8 +76,16 @@ function inject (bot, { version }) {
 
     const villagerPromise = bot.openEntity(villagerEntity)
     bot._client.on(tradeListPacket, gotTrades)
-    const villager = await villagerPromise
+    // [mindaxis-patch:villager-listener-cleanup] Clean up trade_list listener on error
+    let villager
+    try {
+      villager = await villagerPromise
+    } catch (err) {
+      bot._client.removeListener(tradeListPacket, gotTrades)
+      throw err
+    }
     if (villager.type !== 'minecraft:villager' && villager.type !== 'minecraft:merchant') {
+      bot._client.removeListener(tradeListPacket, gotTrades)
       throw new Error('Expected minecraft:villager or minecraft:mechant type, but got ' + villager.type)
     }
 
